workflow.go

```diff
@@ -0,0 +1,257 @@
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Subscription workflow started", "customer", subscription.Customer, "subscription", subscription)
+
+	err = subscription.trialPeriod(ctx)
+	var continueAsNewErr *workflow.ContinueAsNewError
+	if err != nil && errors.As(err, &continueAsNewErr) {
+		return subscription.maybeContinueAsNew(ctx, err)
+	} else if err != nil {
+		logger.Error("Error while handling trial period", "Error", err)
+		return err
+	}
+
+	err = subscription.billingCycle(ctx)
+	if err != nil && errors.As(err, &continueAsNewErr) {
+		return subscription.maybeContinueAsNew(ctx, err)
+	} else if err != nil {
+		logger.Error("Error while handling billing cycle", "Error", err)
+		return err
+	}
+
+	err = subscription.cancelSubscription(ctx)
+	if err != nil {
+		logger.Error("Canceling subscription failed.", "Error", err)
+		return err
+	}
+
+	logger.Info("Subscription workflow completed.")
+	return nil
+}
+
+func (s *SubscriptionInfo) trialPeriod(ctx workflow.Context) (err error) {
+	logger := workflow.GetLogger(ctx)
+	selector := workflow.NewSelector(ctx)
+	info := workflow.GetInfo(ctx)
+
+	var a *Activities
+	var cancelSub bool
+
+	if !s.TrialPeriodExpiration.IsZero() && s.TrialPeriodExpiration.After(workflow.Now(ctx)) {
+		err = workflow.ExecuteActivity(ctx, a.SendWelcomeEmail, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("Welcome Email Activity failed.", "Error", err)
+			return err
+		}
+
+		// Start trial period timer, async
+		trialDuration := s.TrialPeriodExpiration.Sub(workflow.Now(ctx))
+		trialTimer := workflow.NewTimer(ctx, trialDuration)
+
+		var timerError error
+		trialOver := false
+		selector.AddFuture(trialTimer, func(f workflow.Future) {
+			logger.Info("Trial has ended. Sending 'trial is over' email, then entering normal billing cycle.")
+			trialOver = true
+			timerError = f.Get(ctx, nil)
+		})
+
+		for !trialOver {
+			selector.
+				// Signal Handler for updating billing address or other customer info
+				AddReceive(workflow.GetSignalChannel(ctx, "update"), func(c workflow.ReceiveChannel, _ bool) {
+					var updateInfo UpdateSignal
+					c.Receive(ctx, &updateInfo)
+					logger.Info("Received update signal", "data", updateInfo)
+					s = &updateInfo.Subscription
+				}).
+				// Signal handler for canceling the subscription
+				AddReceive(workflow.GetSignalChannel(ctx, "cancel"), func(c workflow.ReceiveChannel, _ bool) {
+					var cancel UpdateSignal
+					c.Receive(ctx, &cancel)
+					logger.Info("Received cancel signal", "data", cancel)
+
+					cancelSub = cancel.CancelSubscription
+				})
+
+			err = workflow.Await(ctx, func() bool {
+				logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+				return info.GetCurrentHistoryLength() > EventHistoryThreshold || selector.HasPending()
+			})
```

Comment:
If the only things coming in are things that affect your selector and you don't send commands inside of select, often it's clearer to read to just check this value after the select call.

EDIT: Now I see how you're draining below, this is actually best. Though the logic can be reworked to have a selector up top.